docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -253,10 +262,86 @@ When a `ClusterResourceSet` is deleted, it will be removed from the `bindings` l
+
+***Detecting changes***
+`ClusterResourceBindings` will contain consistent hash for the resource/s definitions. We will use this to detect changes, comparing the hash of the current resource/s definition with the one stored in the `ClusterResourceBindings`.
+
+Note that this hash will change when any of the resources is updated, a resource is added or a resource is removed. This means that all resources in the same `ConfigMap` or `Secret`, and not only the one that changed, will be reapplied in any of these 3 cases. It also means that resources removed from `ConfigMap` or `Secret` won't be deleted from the target clusters.
+
+In the next *before-after* example we can see that only one resource has changed (`ConfigMap` `calico-configmap`). However, all the 3 resources (`calico-secret1`, `calico-secret2` and `calico-configmap`) will be reapplied.
+
+Before:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: calico-addon
+data:
+  calico1.yaml: |-
+     kind: Secret
+     apiVersion: v1
+     metadata:
+      name: calico-secret1
+      namespace: mysecrets
+      ---
+     kind: Secret
+     apiVersion: v1
+     metadata:
+      name: calico-secret2
+      namespace: mysecrets
+  calico2.yaml: |-
+     kind: ConfigMap
+     apiVersion: v1
+     metadata:
+      name: calico-configmap
+      namespace: myconfigmaps
+     data:
+       key: "original value"
+```
+
+After:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: calico-addon
+data:
+  calico1.yaml: |-
+     kind: Secret
+     apiVersion: v1
+     metadata:
+      name: calico-secret1
+      namespace: mysecrets
+      ---
+     kind: Secret
+     apiVersion: v1
+     metadata:
+      name: calico-secret2
+      namespace: mysecrets
+  calico2.yaml: |-
+     kind: ConfigMap
+     apiVersion: v1
+     metadata:
+      name: calico-configmap
+      namespace: myconfigmaps
+     data:
+       key: "value that changed"
+```
+***Drift***
+
+The proposed solution only deals with changes in the resources' definitions and not with changes in the real objects in the workload clusters. If those objects are modified or deleted in the workload clusters, the `ClusterResourceSet`'s controller won't do anything and they will remain unchanged until their definition in the management cluster is updated.
+
+This could potentially be mitigated by:
+ * Implementing a "periodic" reconciliation mode where resources are reapplied with a certain frequency even their hash hasn't changed.
+ * Storing the compounded `Generation` of the applied objects in the `ResourceSetBinding`. Since `Generation` is a monotonically increasing integer, a change in the compounded generation (adding all the `Generation` fields for all the resources) always means at least one resource changed in the workload cluster. With this mechanism, the hash can be used to detect changes in the resources definition and the compounded generation to detect changes in the actual workload cluster's resource.
```

Comment:
IMO this mitigation doesn't seem to be addressing the drift risk, it instead mitigates possible side effects of "re-apply" all. If this is true, let's have a separated paragraph under risk and mitigations

Reply:
I should have clarified. This will need to be used in conjunction with a periodic check. So the controller will retrieve the objects from the target cluster, calculate the compounded generation and compare it to the one stored in the `ResourceSetBinding`.

Would you prefer if I added that clarification to this bullet point or just put it in a separate paragraph as you proposed?

Reply:
Uhmm, I think that retrying objects from the target cluster could be problematic, given that there are no guarantees that something else - a controller - add something the object spec (thus increasing generation) intentionally.
In this case, we could end up in "deleting" controller changes and this fighting for reconciling the object.
Have you experimented with this idea to get some concrete feedback?

Reply:
No, I haven't. I have experimented with "reapply" when then resource definition changes, but not periodic ones.

This brings up the question of who is the "owner" of an object that is defined by a CRS. IMO the CRS controller is the owner, at least owner of the fields that have been defined in the `ConfigMap` or `Secret`.

As far as deleting controller changes, that's a fair concern. Would using server side apply alleviate it?

Reply:
might be server-side-apply can solve the problem but this requires some thinking, so I'm leaning toward removing this mitigation given that it leads to discussions not core to this proposal.

Reply:
Removed :)